syntax = "proto3";


package FileExchange;

service SimpleFileServer {
    rpc Upload(stream UploadRequest) returns(UploadResponse) {}
    rpc Download(DownloadRequest)    returns(stream DownloadResponse) {}
}

message FileHeader {
    string name = 1;
    uint64 size = 2;
}

message FileChunk {
    bytes data = 1;
}

message UploadRequest {
    oneof contents {
        FileHeader header = 1;
        FileChunk  chunk  = 2;
    }
}

message UploadResponse {
    uint64 bytesReceived = 1;
}

message DownloadRequest {
    string name      = 1;
    uint64 chunkSize = 2;
}

message DownloadResponse {
    oneof contents {
        FileHeader header = 1;
        FileChunk  chunk  = 2;
    }
}

